pkg/session/state.go

```diff
@@ -209,85 +205,105 @@ func (spt *SessionState) RemoveFromRetrieval(storageProviderId peer.ID, retrieva
 	// EMA of connect time
-	if !status.connectTimeInitialized {
-		status.connectTimeInitialized = true
-		status.connectTime = current
+	if !status.connectTimeMs.initialized {
+		status.connectTimeMs.initialized = true
+		status.connectTimeMs.value = currentMs
 	} else {
-		status.connectTime = time.Duration((1-spt.config.ConnectTimeAlpha)*float64(current) + spt.config.ConnectTimeAlpha*float64(status.connectTime))
+		status.connectTimeMs.value = uint64((1-spt.config.ConnectTimeAlpha)*float64(currentMs) + spt.config.ConnectTimeAlpha*float64(status.connectTimeMs.value))
 	}
 	spt.spm[storageProviderId] = status
+
+	if !spt.overallConnectTimeMs.initialized {
+		spt.overallConnectTimeMs.initialized = true
+		spt.overallConnectTimeMs.value = currentMs
+	} else {
+		spt.overallConnectTimeMs.value = uint64((1-spt.config.OverallConnectTimeAlpha)*float64(currentMs) + spt.config.OverallConnectTimeAlpha*float64(spt.overallConnectTimeMs.value))
+	}
 }
 
-func (spt *SessionState) CompareStorageProviders(protocol multicodec.Code, a, b peer.ID, mda, mdb metadata.Protocol) bool {
-	if protocol == multicodec.TransportGraphsyncFilecoinv1 {
-		gsmda := mda.(*metadata.GraphsyncFilecoinV1)
-		gsmdb := mdb.(*metadata.GraphsyncFilecoinV1)
-		// prioritize verified deals over not verified deals
-		if gsmda.VerifiedDeal != gsmdb.VerifiedDeal {
-			return gsmda.VerifiedDeal
-		}
+func (spt *SessionState) CompareStorageProviders(a, b peer.ID, mda, mdb metadata.Protocol) bool {
+	spt.lk.Lock()
+	defer spt.lk.Unlock()
+
+	// score both peers, a float >=0, higher is better
+	gsmda, _ := mda.(*metadata.GraphsyncFilecoinV1)
+	sas := spt.scoreProvider(a, gsmda)
+	gsmdb, _ := mdb.(*metadata.GraphsyncFilecoinV1)
+	sbs := spt.scoreProvider(b, gsmdb)
+	// a's score as a fraction of the total combined score
+	sasf := sas / (sas + sbs)
+	// random dice roll to decide, biasing toward the higher score
+	// DEBUG: fmt.Printf("a(%s)=%f<>b(%s)=%f\n", string(a), sas, string(b), sbs)
+	return spt.config.roll() < sasf
+}
 
-		// prioritize fast retrievel over not fast retrieval
-		if gsmda.FastRetrieval != gsmdb.FastRetrieval {
-			return gsmda.FastRetrieval
+// scoreProvider returns a score for a given provider, higher is better.
+// This method assumes the caller holds lock.
+func (spt *SessionState) scoreProvider(id peer.ID, md *metadata.GraphsyncFilecoinV1) float64 {
+	var score float64
+	// var v, f bool
+
+	// graphsync metadata weighting
+	if md != nil {
+		if md.VerifiedDeal {
+			score += spt.config.VerifiedDealWeight
+			// v = true
+		}
+		if md.FastRetrieval {
+			score += spt.config.FastRetrievalWeight
+			// f = true
 		}
 	}
 
-	spt.lk.Lock()
-	defer spt.lk.Unlock()
-	sa := spt.spm[a]
-	sb := spt.spm[b]
+	// collected metrics scoring
+	sp := spt.spm[id]
+
+	// exponential decay of connect time, normalised by overall connect time
+	// if we have no connect time data, use the average, if we have no average
+	// use 1
+	λ := 1 / float64(spt.overallConnectTimeMs.getValue(1))
```

Comment:
I believe λ here is the rate parameter in an exponential distribution. It might be helpful to provide a link to some references about this in a comment. Especially for folks who want to understand what this symbol means.